x-pack/plugins/lens/public/utils/native_renderer/native_renderer.test.tsx

```diff
@@ -0,0 +1,100 @@
+    const testProps = { a: 'abc' };
+
+    render(<NativeRenderer render={renderSpy} actualProps={testProps} />, mountpoint);
+    render(<NativeRenderer render={renderSpy} actualProps={testProps} />, mountpoint);
+    expect(renderSpy).toHaveBeenCalledTimes(1);
+  });
+
+  it('should not render again if props do not change shallowly', () => {
+    const renderSpy = jest.fn();
+    const testProps = { a: 'abc' };
+
+    render(<NativeRenderer render={renderSpy} actualProps={testProps} />, mountpoint);
+    render(<NativeRenderer render={renderSpy} actualProps={{ ...testProps }} />, mountpoint);
+    expect(renderSpy).toHaveBeenCalledTimes(1);
+  });
+
+  it('should render again once if props change', () => {
+    const renderSpy = jest.fn();
+    const testProps = { a: 'abc' };
+
+    render(<NativeRenderer render={renderSpy} actualProps={testProps} />, mountpoint);
+    render(<NativeRenderer render={renderSpy} actualProps={{ a: 'def' }} />, mountpoint);
+    render(<NativeRenderer render={renderSpy} actualProps={{ a: 'def' }} />, mountpoint);
+    expect(renderSpy).toHaveBeenCalledTimes(2);
+    const containerElement = mountpoint.firstElementChild;
+    expect(renderSpy).lastCalledWith(containerElement, { a: 'def' });
+  });
+
+  it('should render again if props are extended', () => {
+    const renderSpy = jest.fn();
+    const testProps = { a: 'abc' };
+
+    render(<NativeRenderer render={renderSpy} actualProps={testProps} />, mountpoint);
+    render(<NativeRenderer render={renderSpy} actualProps={{ a: 'abc', b: 'def' }} />, mountpoint);
+    expect(renderSpy).toHaveBeenCalledTimes(2);
+    const containerElement = mountpoint.firstElementChild;
+    expect(renderSpy).lastCalledWith(containerElement, { a: 'abc', b: 'def' });
+  });
+
+  it('should render again if props are limited', () => {
+    const renderSpy = jest.fn();
+    const testProps = { a: 'abc', b: 'def' };
+
+    render(<NativeRenderer render={renderSpy} actualProps={testProps} />, mountpoint);
+    render(<NativeRenderer render={renderSpy} actualProps={{ a: 'abc' }} />, mountpoint);
+    expect(renderSpy).toHaveBeenCalledTimes(2);
+    const containerElement = mountpoint.firstElementChild;
+    expect(renderSpy).lastCalledWith(containerElement, { a: 'abc' });
+  });
+
+  it('should render a div as container', () => {
+    const renderSpy = jest.fn();
+    const testProps = { a: 'abc' };
+
+    render(<NativeRenderer render={renderSpy} actualProps={testProps} />, mountpoint);
+    const containerElement: Element = mountpoint.firstElementChild!;
+    expect(containerElement.nodeName).toBe('DIV');
+  });
+
+  it('should render a specified element as container', () => {
+    const renderSpy = jest.fn();
+    const testProps = { a: 'abc' };
+
+    render(<NativeRenderer render={renderSpy} tag="span" actualProps={testProps} />, mountpoint);
+    const containerElement: Element = mountpoint.firstElementChild!;
+    expect(containerElement.nodeName).toBe('SPAN');
+  });
+});
```

Comment:
Can you add a test involving a callback function, like a `setState` function? I would expect a test of the equality of callback functions over multiple renders.

Reply:
Done - is this what you had in mind?